### ISSUE TITLE
Build curl with Android NDK

### ISSUE THREAD

**Issue by MoNTE48** [curl/curl#2058]
I'm trying to compile curl with NDK and get this error.
netrc.c: In function 'Curl_parsenetrc':
CC libcurl_la-rand.lo
netrc.c:81:7: error: implicit declaration of function 'getpwuid_r' [-Werror=implicit-function-declaration]
if(!getpwuid_r(geteuid(), &pw, pwbuf, sizeof(pwbuf), &pw_res)
^

The latest working version is 7.53.1

KDE neon 5.11
NDK R14c

**Comment by MoNTE48** [comment]
https://github.com/curl/curl/issues/1501#issuecomment-342268001
I have this file.

**Comment by jay** [comment]
> I have this file.

Please attach config.log

**Comment by MoNTE48** [comment]
I can not find this file!
Here is the build log
https://gist.github.com/MoNTE48/8fc14fd8fc377ef886580e58098c6606

**Comment by jay** [comment]
> checking for pwd.h... yes
...
> checking for getpwuid_r... yes

Check pwd.h and see if it has a declaration for getpwuid_r

**Comment by MoNTE48** [comment]
@jay, Yes
```
extern int getpwuid_r (__uid_t __uid,
		       struct passwd *__restrict __resultbuf,
		       char *__restrict __buffer, size_t __buflen,
		       struct passwd **__restrict __result)
		       __nonnull ((2, 3, 5));
```

**Comment by jay** [comment]
My guess is it has something to do with compiler flags or cross compiling. Likely your getpwuid_r is guarded by a posix macro and something is preventing the macro from being defined.

**Comment by bagder** [comment]
You should be able to tell your compiler to just ignore this warning/error - if it truly only is a problem with the include file.

**Comment by MoNTE48** [comment]
I will be grateful if you tell me where the error is, or which flag I should use
https://github.com/MultiCraftProject/MultiCraft/blob/master/build/android/Makefile#L186

**Comment by jay** [comment]
Ok. I don't see anything in the CFLAGS that would do that. Turn on verbose so you can see the gcc command that builds netrc.c, `make V=1`. that may help. What you are looking for is ansi which disables posix I believe. Also see https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html

**Comment by MoNTE48** [comment]
https://gist.github.com/MoNTE48/7ef4a0638e864640cf0a853919019c0c

**Comment by jay** [comment]
I didn't see anything in the flags that could cause it. Maybe it is default behavior. Try using that function independently of curl and see if you still get the warning.

**Comment by MoNTE48** [comment]
@jay I'm sorry, I'm not too strong in linux. Can you tell what command I should perform?

**Comment by jay** [comment]
I meant you could try like below and see if it still gives the warning
~~~c
cat << EOF > a.c
#include <pwd.h>
#include <unistd.h>
#include <sys/types.h>
int main() {
struct passwd pw, *pw_res;
char pwbuf[1024];
return !getpwuid_r(geteuid(), &pw, pwbuf, sizeof(pwbuf), &pw_res);
}
EOF
arm-linux-androideabi-gcc a.c 
~~~

**Comment by bagder** [comment]
No response and really looks like a toolchain problem. Closing.

**Comment by pmogilevskiy** [comment]
Hi there, I would like to raise this issue. I have the same issue building the current master on Android with NDK 13b, API level 19. I opened pwd.h and found that getpwuid_r function is disabled by macro with comment: MISSING FROM BIONIC. But, according to the config.log, getpwuid_r exists. Please look at the attached log.
[config.log](https://github.com/curl/curl/files/1598476/config.log)

I investigated this issue a bit and found the commit which breaks the build for me.  It's 9168e2470d5203661faa5500361b7eb493aa24dd

--disable-werror configure option doesn't work

Reverting 9168e2470d5203661faa5500361b7eb493aa24dd fixes the problem.

Could you review 9168e2470d5203661faa5500361b7eb493aa24dd please?

Please let me know if you need more info.

Thanks!

**Comment by MarcelRaad** [comment]
@pmogilevskiy How exactly is `getpwuid_r` disabled in pwd.h? After you revert the `-Werror-implicit-function-declaration` commit, does the resulting program actually run on an API level 19 (Android 4.4) device?

**Comment by pmogilevskiy** [comment]
@MarcelRaad , it disabled by #if 0 Please find pwd.txt in the attachment.
[pwd.txt](https://github.com/curl/curl/files/1598825/pwd.txt)

I didn't try to run on Android 4.4 because I don't have such device. But I tried to run on Android 8.1.0 and Android 5.1.1. It works well.

**Comment by pmogilevskiy** [comment]
Have you had a chance to look at this issue? Do you need more info?

**Comment by MarcelRaad** [comment]
@pmogilevskiy Not yet, sorry. I'm away from keyboard over the weekend.

**Comment by MarcelRaad** [comment]
@pmogilevskiy That's a bug in the NDK. It's fixed in the API level 21 (Android 5.0) header in your NDK 13b, and it's also fixed in the unified headers shipped with NDK version 14 and later. We could implement a workaround for that bug, but my autoconf skills are almost nonexisting yet.

**Comment by pmogilevskiy** [comment]
Thanks @MarcelRaad , Unfortunately, we don't have a chance to use API greater than 19. We'll consider to upgrade NDK to 14. For now, we are good with reverting the commit which I mentioned above in this thread.
